### ISSUE TITLE
RFE: Option for VNC console in separate browser window

### ISSUE THREAD

**Issue by dowdle** [cockpit-project/cockpit-machines#965]
I applaud cockpit-machines for offering an embedded VNC display... but so many other projects are providing a separate window with a full-screen option via noVNC.  I would love to have a pleasingly usable desktop experience.

**Comment by garrett** [comment]
Thanks for your feedback! I agree, we should have this.

Right now, we do have the "expand" button, which gives you a bit more space, but it's not quite the same. It should have even more space. However, middle clicking on the link doesn't work, so you can't open it in a new window.

Also we do have the ability to open up a native app, but it does need improving too (ports and firewall might get in the way or need more configuring, however). One of the blockers is the lack for an ssh option in the `.vv` file. It seems like an omission upstream, so we're waiting on that to be fixed to be able to support it. That's an issue with virt-viewer and it's filed @ https://gitlab.com/virt-viewer/virt-viewer/-/issues/86. More information about the native app support (including this and much more) is on the issue @ https://github.com/cockpit-project/cockpit-machines/issues/553.

Meanwhile, I'll also try to raise the priority of us working on the VNC component, so we can hopefully work on a fix for this sooner.

**Comment by garrett** [comment]
I was considering closing this as a duplicate of https://github.com/cockpit-project/cockpit-machines/issues/680, but I think this captures an idea that the other one doesn't:

- opening the expanded view in a separate tab currently isn't even possible

We should _at least_ start by making middle clicks and control clicks work, to make the VNC viewer work in another tab (or window).

**Comment by dowdle** [comment]
garrett,

One thing you completely side-stepped was my mention of noVNC.  noVNC is packaged in Fedora and is used by several dozen other projects like Proxmox VE, TrueNAS Scale, well... pretty much any virtualization related product that offers a web-based interface.  I'm hoping that there isn't anything in Cockpit's design that would hinder noVNC integration... but I'm not a developer myself.  While the integration might take some effort, all of the features that noVNC comes with would mean, I assume, less effort in having to create those from scratch.

Anyhoo, something to consider.  Having a small region that in the existing window is really a non-starter.

Thanks.

**Comment by garrett** [comment]
@dowdle: Cockpit-Machines actually uses noVNC, as Cockpit-Machines uses PatternFly React Console, which in turn uses noVNC:
https://www.npmjs.com/package/@patternfly/react-console?activeTab=dependencies

When you said "desktop experience", I thought you meant a desktop VNC client.

We already have a separate issue for fullscreen @:

- https://github.com/cockpit-project/cockpit-machines/issues/680

> Having a small region that in the existing window is really a non-starter.

After you click expand, it shouldn't be so small. But it doesn't properly size vertically:

![Screenshot 2023-03-13 at 12-49-52 Virtual machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/224695928-0a7b5ba0-df18-4c3b-a2bb-027bb5192dee.png)

(Fellow devs who working on the VNC viewer: See all that blank space below the VNC screen? It's not being used, and to preserve the aspect ratio, the viewer is adding unnecessary bars on the sides. :frowning_face:)

We already have some issues for having an even larger view (not just a smaller view for Cockpit itself, with the nav hidden @ https://github.com/cockpit-project/cockpit/issues/18433 — possibly even proper fullscreen for VNC, if possible) and intend on working on it soon. It's open as issues here on GitHub (with some designs), some of the team has been working on VNC related issues lately, and it's even on our Jira tracking board as a larger scoped project to-be-done soon.

In other words: We'll hopefully have it solved soon?

**Comment by guercheLE** [comment]
I wouldn't mind having the expand button letting the viewer take all the tab canvas, and if browser is in full screen mode, the viewer taking all the screen, but keeping a button somewhere to collapse it.

**Comment by hw-claudio** [comment]
A separate detachable VNC window is much more usable. Any news on this?

**Comment by garrett** [comment]
Same state as my [comment above](https://github.com/cockpit-project/cockpit-machines/issues/965#issuecomment-1466023931). It's something we want to implement, but we still need a developer to work on this.

**Comment by mvollmer** [comment]
Everybody, please have a look at this demo: https://www.youtube.com/watch?v=FsBE4eQo0As

(Watch it in fullscreen! :-)

We are ready and eager to implement some fullscreen goodness, but need feedback on how it should look exactly.

Some questions we are struggling with:

 - Is it OK to eat the Escape key while in full-monitor-mode? We could do the same thing as Youtube does to go fullscreen, but the browser would intercept the Escape key. Would that be a problem? (We can add it into the "Send keys" menu, but vim users will likely go mad.)

 - In addition to a full-monitor-like-youtube mode, we would also provide a "full-browser-window" mode (like the Youtube Cinema mode). This would replace the existing "Expand" functionality. The question: Should we keep it as a "expand/collapse" toggle operation, or should we have a "Open in new window" button.

 - How would you feel if instead of the "full-browser-window" mode we would allow the navigation sidebar to be toggled in and out? So clicking "Expand" would fill the content area like it does now, and if you need a bigger console still, you have to toggle away the sidebar explicitly.

Thanks!

**Comment by hw-claudio** [comment]
very promising demo!
1) Not sure about the escape key, it's quite used I think, so I would personally prefer it not to be intercepted. I don't think there is a lot of value if jumping in and out of fullscreen mode
2) I would be more in favor of opening in a separate window or tab, rather than the expand/collapse thing. Keep the minimal console there, but when one wants he can click and detach the console into a responsive, separate window, normal or fullscreen.

3) When someone is actually working inside the VM, there is absolutely no need to use anything cockpit related. Any controls present in the window should relate to the VM / console only.

I think a good guideline is the web version of the vmware vsphere web console.

**Comment by hw-claudio** [comment]
congrats on the demo, btw!!!

**Comment by dowdle** [comment]
I liked what I saw in your linked video. I do not mind it eating the ESC key... as I'm used to hitting ESC to get out of full screen mode. I'm not too picky as long as you give me a full screen mode.

I like the way Proxmox VE does it although I'm not positive if their little controls pop-out menu in the center left is needed or not. It would be nice to have a scale option... so for example, when you expanded and your text console went full-screen, it pretty much stayed about the same size it was... and just got a ballpark of borders surrounding it. Imagine a way to have the display actually scale to fit that area, eliminating the borders... while also retaining the proper aspect ratio... so depending on the resolution of the display, there might still be a border in one direction.

I'm excited to see what you come up with... but in the mean time, would love to have what you showed in the video.

TYL,
Scott Dowdle
(406) 750-3319 [cell](406) 994-3931 [work]

**Comment by tgwaste** [comment]
This looks promising but you can not override the Esc Key.  Some of us have been using Escape to exit input mode in VI for like 30 years.
A pop-out that is exactly the size of the usable window would be much better and much more professional. Having all that dead space just doesn't look good.  It works sure, but it should be slick to. ;)
A good example is how VULTR does it with their VMs.

**Comment by dowdle** [comment]
Correct me if I'm wrong, but ESC only gets trapped when the window is in full-screen mode. I don't think you are going to be using vi when you can even see it. Not sure how that plays into multiple screens... but if the browser window isn't the active window, I don't think it'll trap the keys.

TYL,
Scott Dowdle
(406) 750-3319 [cell](406) 994-3931 [work]

**Comment by tgwaste** [comment]
> Correct me if I'm wrong, but ESC only gets trapped when the window is in full-screen mode. I don't think you are going to be using vi when you can even see it. Not sure how that plays into multiple screens... but if the browser window isn't the active window, I don't think it'll trap the keys.
> 
> TYL,
> Scott Dowdle
> (406) 750-3319 [cell](406) 994-3931 [work]
> […](#)

I don't know what any of this means. Did you mean "canT see it"?  Why would I not be able to see something in full screen mode?

Don't trap keys, and don't presume to know what the user is going to want to do or not.

**Comment by mvollmer** [comment]
Alright, I have explored the "Detach" behavior a bit: https://youtu.be/eqAaXCJc3e4

I am quite happy with it, not least because we don't need to change the Shell for that.

@tgwaste, thanks for the suggestion to make it slick and resize the browser window automatically! My instincts would have been not to bother. :-)

**Comment by mvollmer** [comment]
And another one: https://youtu.be/Z9RyJn9f1go

Now we keep "Expand/Collapse", which is probably what you want for Serial.

**Comment by tgwaste** [comment]
> Alright, I have explored the "Detach" behavior a bit: https://youtu.be/eqAaXCJc3e4
> 
> I am quite happy with it, not least because we don't need to change the Shell for that.
> 
> [@tgwaste](https://github.com/tgwaste), thanks for the suggestion to make it slick and resize the browser window automatically! My instincts would have been not to bother. :-)

There ya go!  That's looking a lot better!

**Comment by hw-claudio** [comment]
agreed, this is on the right track!

**Comment by tgwaste** [comment]
hi,  this releasing soon?  .. or should it already be?

**Comment by dowdle** [comment]
cockpit-machines-337 is current (in Fedora 42). The update will be in 338 which you can find on Koji (a single package download) if you want it before it hits updates. I don't know if it has been submitted to updates yet or not.

I installed the koji package and it worked just fine.

TYL,
Scott Dowdle
(406) 750-3319 [cell](406) 994-3931 [work]

**Comment by martinpitt** [comment]
https://bodhi.fedoraproject.org/updates/FEDORA-2025-0385d64463 landed in stable, so F42 has it now. F41 is on its way in, and F43 is stuck in freeze.
